dev/options.go

```diff
@@ -0,0 +1,101 @@
+package dev
+
+import (
+	"io"
+	"time"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+type WithLogger logr.Logger
+
+func (l WithLogger) ApplyToEnvironmentConfig(c *EnvironmentConfig) {
+	c.Logger = logr.Logger(l)
+}
+
+func (l WithLogger) ApplyToClusterConfig(c *ClusterConfig) {
+	c.Logger = logr.Logger(l)
+}
+
+func (l WithLogger) ApplyToWaiterConfig(c *WaiterConfig) {
+	c.Logger = logr.Logger(l)
+}
+
+type WithInterval time.Duration
+
+func (i WithInterval) ApplyToWaiterConfig(c *WaiterConfig) {
+	c.Interval = time.Duration(i)
+}
+
+type WithTimeout time.Duration
+
+func (t WithTimeout) ApplyToWaiterConfig(c *WaiterConfig) {
+	c.Timeout = time.Duration(t)
+}
+
+type WithSchemeBuilder runtime.SchemeBuilder
+
+func (sb WithSchemeBuilder) ApplyToClusterConfig(c *ClusterConfig) {
+	c.SchemeBuilder = runtime.SchemeBuilder(sb)
+}
+
+type WithNewWaiterFunc NewWaiterFunc
+
+func (f WithNewWaiterFunc) ApplyToClusterConfig(c *ClusterConfig) {
+	c.NewWaiter = NewWaiterFunc(f)
+}
+
+type WithWaitOptions []WaitOption
+
+func (opts WithWaitOptions) ApplyToClusterConfig(c *ClusterConfig) {
+	c.WaitOptions = []WaitOption(opts)
+}
+
+type WithNewHelmFunc NewHelmFunc
+
+func (f WithNewHelmFunc) ApplyToClusterConfig(c *ClusterConfig) {
+	c.NewHelm = NewHelmFunc(f)
+}
+
+type WithHelmOptions []HelmOption
+
+func (opts WithHelmOptions) ApplyToClusterConfig(c *ClusterConfig) {
+	c.HelmOptions = []HelmOption(opts)
+}
+
+type WithStdout struct{ io.Writer }
+
+func (w WithStdout) ApplyToHelmConfig(c *HelmConfig) {
+	c.Stdout = io.Writer(w)
+}
+
+type WithStderr struct{ io.Writer }
+
+func (w WithStderr) ApplyToHelmConfig(c *HelmConfig) {
+	c.Stderr = io.Writer(w)
+}
+
+type WithClusterInitializers []ClusterInitializer
+
+func (i WithClusterInitializers) ApplyToEnvironmentConfig(c *EnvironmentConfig) {
+	c.ClusterInitializers = append(c.ClusterInitializers, i...)
+}
+
+type WithContainerRuntime string
```

Comment:
I have added consts for `Podman` and `Docker`, but this still allows people to use any string as well, because it's a string type cast...

Other options:
- make `WithContainerRuntime` a struct, so it's not a type cast
- split `WithContainerRuntime` into `WithDocker`/`WithPodman`, which would lead to ambiguity if both are set.

Go does not have real enums and I don't use `iota`, as it's _very_ easy to screw something up with iota in public APIs.
Also having an `int` instead of a `string` constant is not really helping, as you can still type cast any int into the option type...

The only real solution is to have validation in place, but that is IMHO out of the scope of this PR...

Reply:
That's more than sufficient. The idea was more to hint at accepted values rather than have someone type `Podman` or `podma` and get frustrated. At least this way the compiler helps lead the way.